[HPB-152] [BE] 登入登出 API

richardlinio opened the pull request:
[Issue Link](https://www.notion.so/BE-API-276cba948f1180958503e06a6662ec29?source=copy_link)

本 PR 需與以下 PR 一起 review：https://github.com/Goodidea-backend-camp/hpb-blog-project/pull/9

## Summary

本 PR 為 HPB Blog 專案實作完整的用戶認證系統，包含：
- JWT 驗證機制
- 登入/登出 API 端點
- 資安防護（防時序攻擊、bcrypt 密碼加密、security header and host header validation middleware）

## Major Changes

整體架構如下：

```
┌─────────────────────────────────────────────────────────────┐
│                         HTTP 請求                            │
└──────────────────────┬──────────────────────────────────────┘
                       │
                       ▼
┌─────────────────────────────────────────────────────────────┐
│                    Middleware 層                             │
│  - CORS (允許的來源驗證)                                       │
│  - SecurityHeaders (安全 HTTP headers)                       │
│  - HostHeaderValidation (Host header 驗證)                   │
│  - AuthRequired (JWT 令牌驗證，保護特定路由)                    │
└──────────────────────┬──────────────────────────────────────┘
                       │
                       ▼
┌─────────────────────────────────────────────────────────────┐
│              AuthHandler (HTTP API 層)                       │
│  - POST /auth/login   → Login()                              │
│  - POST /auth/logout  → Logout() [需 AuthRequired]           │
└──────────────────────┬──────────────────────────────────────┘
                       │
                       ▼
┌─────────────────────────────────────────────────────────────┐
│           AuthRepository (資料存取層)                         │
│  - GetUserByUsername()                                      │
│  - 領域錯誤處理 (ErrUserNotFound, ErrTimeout 等)               │
└──────────────────────┬──────────────────────────────────────┘
                       │
                       ▼
┌─────────────────────────────────────────────────────────────┐
│              Database (PostgreSQL)                          │
│  - users 表查詢                                              │
└─────────────────────────────────────────────────────────────┘

                       ┌─────────────────┐
                       │   JWT Utils     │
                       │  (auth package) │
                       │ - GenerateToken │
                       │ - ValidateToken │
                       │ - ValidateSecret│
                       └─────────────────┘
```

### 1. 認證核心功能
- **JWT 工具模組** ([internal/auth/jwt.go](internal/auth/jwt.go))
  - 實作 JWT 生成與驗證
  - 使用 HS256 簽名算法
  - token 有效期 7 天
  - 包含 user ID 和標準 Claims

### 2. 資料存取層 (Repository)
- **AuthRepository** ([internal/repository/auth.go](internal/repository/auth.go))
  - 實作用戶查詢介面 `AuthRepository`
  - 定義 `AuthQuerier` 介面，解耦資料庫查詢邏輯
  - 實作 `GetUserByUsername` 方法根據用戶名查詢用戶資料
  - 支援軟刪除過濾
  - **Domain 錯誤處理**：定義領域層錯誤類型
    - `ErrUserNotFound`：用戶不存在
    - `ErrDatabaseQuery`：資料庫查詢錯誤
    - `ErrDatabaseConn`：資料庫連線錯誤
    - `ErrTimeout`：查詢超時
    - `ErrCanceled`：查詢被取消

### 3. HTTP API 層
- **AuthHandler** ([internal/api/auth.go](internal/api/auth.go))
  - `POST /auth/login` - 用戶登入
    - 驗證用戶名和密碼
    - 返回 JWT 令牌和用戶資訊
    - **查詢超時控制**：使用 `context.WithTimeout()` 設定 5 秒超時
    - **安全特性**：防止時序攻擊，無論用戶是否存在都執行相同的密碼驗證流程（使用 dummy password hash）
    - **錯誤處理**：處理 `repository.ErrUserNotFound` 而非直接使用 `pgx.ErrNoRows`
    - **可配置性**：支援透過 `WithBcryptCost()` 選項配置 bcrypt 成本係數
    - 失敗時返回 **401 Unauthorized**
  - `POST /auth/logout` - 用戶登出
    - 由 AuthRequired middleware 驗證 JWT 有效性
    - 返回 204 No Content，實際登出由 client side 處理

### 4. Middleware 層
- **AuthRequired** ([internal/middleware/auth.go](internal/middleware/auth.go))
  - 從 `Authorization` header 提取 Bearer token
  - 驗證 token 有效性並解析用戶 ID
  - 將 `user_id` 設置到 context 中供後續 handler 使用
  - 認證失敗返回 **401 Unauthorized**

- **SecurityHeaders** ([internal/middleware/security.go](internal/middleware/security.go))
  - 設置安全相關的 HTTP headers
  - `X-Content-Type-Options: nosniff`
  - `X-Frame-Options: DENY`
  - `X-XSS-Protection: 1; mode=block`
  - `Strict-Transport-Security: max-age=31536000; includeSubDomains`
  - `Content-Security-Policy: default-src 'self'`

- **HostHeaderValidation** ([internal/middleware/security.go](internal/middleware/security.go))
  - 驗證 HTTP `Host` header，防止 Host header 注入攻擊
  - 支援多個允許的後端主機（逗號分隔）
  - 從環境變數 `ALLOWED_BACKEND_HOSTS` 讀取允許的主機列表
  - 驗證失敗返回 **400 Bad Request**

### 5. 伺服器整合
- **主程式** ([cmd/server/main.go](cmd/server/main.go))
  - 初始化 `AuthRepository` (取代 AuthStore)
  - **JWT 密鑰驗證**：使用 `auth.ValidateSecret()` 確保密鑰長度至少 32 字元
  - 從環境變數讀取配置
  - **CORS middleware**：使用 `gin-contrib/cors`，從 `ALLOWED_ORIGINS` 讀取允許的來源
  - **security middleware**：應用 `SecurityHeaders()` 和 `HostHeaderValidation()`
  - 註冊認證路由到 Gin Router 的 `/auth` 群組

### 6. 開發工具
- **測試用戶 seed script** ([scripts/seed.go](scripts/seed.go))
  - 預設生成用戶：`username: "test"`, `password: "test"`
  - 自動使用 bcrypt 加密密碼

- **Makefile 命令** ([Makefile](Makefile))
  - `make seed` - 在 Docker 容器中執行種子腳本，自動插入或更新測試用戶到資料庫


### 7.安全特性

1. **防時序攻擊**：登入時無論用戶是否存在，都執行相同的 bcrypt 驗證，確保響應時間一致
2. **密碼加密**：使用 bcrypt 加密密碼，預設成本係數 10
3. **JWT 簽名**：使用 HS256 算法簽名，防止令牌篡改
4. **JWT 密鑰強度**：強制 JWT 密鑰長度至少 32 字元
5. **Token 過期**：JWT  7 天後自動過期
6. **統一錯誤訊息**：登入失敗時返回統一訊息，避免洩露用戶是否存在
7. **查詢超時控制**：資料庫查詢設定 5 秒超時，防止長時間阻塞
8. **安全 Headers**：設定多層安全 HTTP headers (CSP, XSS Protection, Frame Options 等)
9. **Host Header 驗證**：防止 Host header 注入攻擊
10. **CORS 保護**：限制跨域請求來源

## How to Test

### 1. 環境準備
1. 切換到 hpb-blog-project 資料夾
2. fetch 並切換到 feature/HPB-152 分支
3. `cp .env.dev .env`，確保 .env 環境變數已設定：
```bash
# Auth
JWT_SECRET=jwt-secret-key-for-dev-only-please-change-in-production

# CORS Configuration
ALLOWED_ORIGINS=http://localhost:3000,https://www.hpb.testdata.work,https://admin.hpb.testdata.work

# Host Header Validation (允許多個值，用逗號分隔)
ALLOWED_BACKEND_HOSTS=api.hpb.testdata.work,backend:8080
```

4. `docker compose up --build` 確認無錯誤

### 2. 生成測試用戶
1. 切換到 hpb-blog-backend 資料夾
2. fetch 並切換到 feature/HPB-152 分支
3. `make seed`，這會創建或更新一組測試用的帳號密碼：`username: "test"`, `password: "test"`，可以在資料庫中看到這筆資料（密碼為 hash）

### 3. 測試登入 API
1. 使用 postman 測試 https://api.hpb.testdata.work/auth/login POST 請求
2. body 不輸入任何資料，應回傳 400
<img width="1394" height="1060" alt="Screenshot 2025-11-16 at 1 49 36 PM" src="https://github.com/user-attachments/assets/bf865360-ce57-49b0-beb1-e8d6db1c9f75" />

4. body 輸入錯誤帳號密碼，應回傳 401
<img width="1080" height="876" alt="Screenshot 2025-11-16 at 12 06 50 AM" src="https://github.com/user-attachments/assets/b98e346a-bb5a-4b83-bf34-09913e0c706b" />

5. 使用以下 body 登入，應獲得 jwt token 與 username
```json
{
  "username": "test",
  "password": "test"
}
```
<img width="1062" height="1034" alt="Screenshot 2025-11-16 at 12 07 20 AM" src="https://github.com/user-attachments/assets/9a9b05a3-b9fb-4fbf-ba49-205457cd4121" />
4. 複製 jwt token 備用

### 4. 測試登出 API
1. 不使用 jwt token，用 postman 測試 https://api.hpb.testdata.work/auth/logout POST 請求，應回傳 401
<img width="1066" height="850" alt="Screenshot 2025-11-16 at 12 09 30 AM" src="https://github.com/user-attachments/assets/01a5e9ae-fb45-47de-ba00-ec8179bcfbb7" />

2. 點擊 Authorization 標籤，Auth Type 選擇 Bearer token，輸入剛剛的 jwt token

3. 此時用 postman 測試 https://api.hpb.testdata.work/auth/logout POST 請求，應回傳 204

<img width="1062" height="744" alt="Screenshot 2025-11-16 at 12 07 57 AM" src="https://github.com/user-attachments/assets/49f6e207-d16e-48bb-8873-db1c60961546" />


## Known Issues

- [[HPB-210] [BE] 後端 rate limiting middleware](https://www.notion.so/jyu1999/BE-rate-limiting-middleware-2a9cba948f1180b3bda9cd8591f59804?source=copy_link)
- [[HPB-211] [BE] 以 logger 記錄 server error 以利 debug](https://www.notion.so/jyu1999/BE-logger-server-error-debug-2a9cba948f1180458b2dc2c1713e17b1?source=copy_link)
- [[HPB-212] [BE] Auth 機制由 JWT 遷移至 Access token + Refresh token](https://www.notion.so/jyu1999/BE-Auth-JWT-Access-token-Refresh-token-2a9cba948f118075876df3b83577d7c7?source=copy_link)
- [[HPB-213] [BE] 增加使用者帳號密碼驗證規則](https://www.notion.so/jyu1999/BE-2a9cba948f1180edb81ac8d55a66271b?source=copy_link)
- [[HPB-217] [BE] 把過期或已登出的 token 加入黑名單，登入前檢查 token 是否在黑名單上](https://www.notion.so/jyu1999/BE-token-2accba948f1180dcbc21c1b301d229a1?source=copy_link)

richardlinio committed:
feat(db): add query and method to get user by username
- add SQL query GetUserByUsername to select user by username if not deleted
- generate db method GetUserByUsername to fetch user by username from database
- update Querier interface to include GetUserByUsername method signature

richardlinio committed:
feat(auth): add JWT authentication support

- add github.com/golang-jwt/jwt/v5 dependency
- define JWTExpiration constant for token expiry duration
- implement JWT token generation and validation functions with custom Claims
- handle errors for invalid token, expired token, and missing secret key
- add comprehensive tests for token generation, validation, expiration, invalid tokens, and secret key absence

richardlinio committed:
feat(store): add AuthStore interface and implementation for user authentication
- define AuthQuerier interface for database user retrieval
- implement authStore struct with GetUserByUsername method
- provide NewAuthStore constructor for creating AuthStore instances
- add comprehensive unit tests covering success, error cases, context cancellation, and concurrency scenarios

richardlinio committed:
feat(auth): add authentication handler with login and logout endpoints
- Introduce AuthHandler for handling login/logout logic with JWT support
- Implement login with timing attack prevention and bcrypt password verification
- Implement logout endpoint validating JWT token without server-side token storage
- Add environment variable JWT_SECRET validation in server main initialization
- Register authentication routes under `/auth` group in Gin router
- Provide comprehensive unit tests covering success, failure, edge cases, concurrency, and security aspects

richardlinio committed:
fix(auth): add JWT secret validation and enforce minimum length

- add ValidateSecret function to check JWT secret is non-empty and at least 32 characters
- update main.go to validate JWT secret on startup with error wrapping
- define new error ErrSecretTooShort for short secrets
- add comprehensive tests for ValidateSecret including boundary, unicode, special chars, and concurrency scenarios

richardlinio committed:
feat(seed): replace factory user generation with database seeding command

- Remove cmd/factory with its user generation and SQL output logic
- Add cmd/seed to directly seed test users into the database with hashed passwords
- Update Makefile to add seed target and remove factory and factory-seed targets

richardlinio committed:
perf(api): add database query timeout for login to prevent indefinite blocking
- define loginTimeout constant for maximum login DB operation duration
- use context.WithTimeout with loginTimeout in Login handler when calling GetUserByUsername
- add unit test to verify database query timeout triggers proper error response

richardlinio committed:
feat(auth): add ExtractBearerToken to parse and validate Bearer token from Authorization header
- Introduce ExtractBearerToken function to extract JWT token with validation of header presence, prefix, and token emptiness
- Add related error types for missing header, invalid format, and empty token
- Update JWT package imports and constants accordingly

feat(middleware): add AuthMiddleware for JWT authentication validation
- Implement middleware to extract and validate JWT token from Authorization header
- Abort request with 401 Unauthorized and error response on invalid/missing token
- Set user_id in Gin context on successful validation for downstream handlers

refactor(api/auth_handler): protect logout route with AuthMiddleware and simplify Logout handler
- Apply AuthMiddleware to /auth/logout endpoint to enforce authentication
- Remove token extraction and validation from Logout handler as it is now middleware responsibility
- Update Logout handler to simply return 204 No Content on success

fix(api/auth_handler_test): update logout tests to expect 401 Unauthorized instead of 403 Forbidden
- Change expected HTTP status codes and error messages in logout tests to align with middleware error handling
- Adjust error message expectations for invalid authorization formats and tokens

test(auth): add tests for ExtractBearerToken function
- Cover valid token extraction, missing header, wrong prefix, empty token, and edge cases like long tokens and unicode
- Include concurrent tests to ensure thread safety

test(middleware): add tests for AuthMiddleware
- Test valid tokens pass and user_id is set in context
- Test missing or malformed Authorization headers result in 401 Unauthorized with proper error messages
- Test expired and invalid tokens fail authentication correctly
- Verify context propagation of user_id to downstream handlers

chore(go.mod): add testify and go-spew dependencies for testing and debugging
- Add github.com/stretchr/testify for assertions in tests
- Add github.com/davecgh/go-spew and github.com/pmezard/go-difflib as indirect dependencies for test utilities

richardlinio committed:
refactor(api, store): rename auth files for consistency
- Rename auth_handler.go to auth.go
- Rename auth_handler_test.go to auth_test.go
- Rename auth_store.go to auth.go
- Rename auth_store_test.go to auth_test.go

richardlinio committed:
feat(middleware): add security middlewares and CORS configuration

- Add CORS middleware with allowed origins, methods, headers, credentials, and max age support
- Introduce SecurityHeaders middleware to set various HTTP security headers
- Implement HostHeaderValidation middleware to validate Host header and prevent SSRF attacks
- Enforce BACKEND_DOMAIN and ALLOWED_ORIGINS environment variables for security configuration
- Integrate security middlewares into Gin router initialization in main.go
- Add comprehensive tests for SecurityHeaders and HostHeaderValidation middlewares
- Upgrade dependencies including gin-contrib/cors and update indirect packages

richardlinio committed:
refactor(auth): rename AuthMiddleware to AuthRequired for clarity

- Rename AuthMiddleware to AuthRequired in middleware/auth.go and related comments
- Update all references from AuthMiddleware to AuthRequired in internal/api/auth.go
- Rename test functions and update middleware usage in internal/middleware/auth_test.go accordingly

richardlinio committed:
refactor(api): move auth-related request and response structs to top of auth.go
- relocate LoginRequest, LoginUser, LoginResponse, and ErrorResponse type definitions from bottom to top of file
- group all auth-related data structures together for better code organization and readability

richardlinio committed:
refactor(seed): move seed script to scripts directory and update Makefile

- Rename seed script from cmd/seed/main.go to scripts/seed.go
- Update Makefile seed target to run the new script path

richardlinio committed:
refactor(auth): replace store.AuthStore with repository.AuthRepository abstraction
- move AuthStore implementation from internal/store to internal/repository as AuthRepository
- rename types and variables from AuthStore to AuthRepository accordingly
- update cmd/server and internal/api/auth.go to use AuthRepository instead of AuthStore
- rename and update tests from internal/store/auth_test.go to internal/repository/auth_test.go
- update all tests in auth_test.go to use mockAuthRepository instead of mockAuthStore

richardlinio committed:
refactor(auth): unify error handling for user not found in auth repository and handler
- Introduced domain error ErrUserNotFound in repository package to replace pgx.ErrNoRows
- Updated AuthRepository.GetUserByUsername to map pgx.ErrNoRows to ErrUserNotFound and wrap other DB errors
- Modified AuthHandler.Login to check for repository.ErrUserNotFound instead of pgx.ErrNoRows for user existence
- Updated tests to expect repository.ErrUserNotFound instead of pgx.ErrNoRows
- Removed unused pgx import in auth.go and auth_test.go
- Added TODO comment for logging internal errors in Login handler

richardlinio committed:
refactor(middleware): support multiple allowed backend hosts for HostHeaderValidation
- rename environment variable from BACKEND_DOMAIN to ALLOWED_BACKEND_HOSTS and parse as comma-separated list
- update HostHeaderValidation middleware to accept slice of allowed hosts instead of single host string
- modify middleware logic to validate Host header against multiple allowed hosts
- update server main.go to set middleware with multiple allowed hosts
- adjust security middleware tests to use slice of allowed hosts accordingly

richardlinio committed:
fix(auth): change invalid login response status code to 401 Unauthorized
- update HTTP response status from 403 Forbidden to 401 Unauthorized in Login handler for invalid username or password
- update corresponding test cases to expect 401 Unauthorized instead of 403 Forbidden on authentication failures and SQL injection attempts

richardlinio committed:
refactor(server): use constant for CORS max age duration
- define corsMaxAge constant for 12 hours
- replace inline MaxAge value with corsMaxAge constant in CORS config

richardlinio committed:
fix(cmd/server): enforce BACKEND_PORT environment variable requirement
- add error for missing BACKEND_PORT variable
- change default BACKEND_PORT fallback to error return instead of defaulting to "8080"

richardlinio committed:
perf(server): increase CORS max age duration
- Change CORS max age from 12 hours to 24 hours to improve caching and reduce preflight requests

richardlinio commented on internal/api/auth.go:
```diff
@@ -0,0 +1,185 @@
+	dummyPasswordHash string
+}
+
+// AuthHandlerOption is a function that configures an AuthHandler.
+type AuthHandlerOption func(*authHandlerConfig)
+
+// authHandlerConfig holds configuration for creating an AuthHandler.
+type authHandlerConfig struct {
+	bcryptCost int
+}
+
+// WithBcryptCost sets a custom bcrypt cost for the AuthHandler.
+// This is primarily used for testing to speed up tests by using bcrypt.MinCost.
+func WithBcryptCost(cost int) AuthHandlerOption {
+	return func(cfg *authHandlerConfig) {
+		cfg.bcryptCost = cost
+	}
+}
+
+// NewAuthHandler creates a new AuthHandler instance.
+func NewAuthHandler(authRepo repository.AuthRepository, jwtSecret string, opts ...AuthHandlerOption) (*AuthHandler, error) {
+	if jwtSecret == "" {
+		return nil, errors.New("JWT secret cannot be empty")
+	}
+
+	// Apply default configuration
+	cfg := &authHandlerConfig{
+		bcryptCost: bcrypt.DefaultCost,
+	}
+
+	// Apply provided options
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
```
這裡的寫法，目的是為了讓 AuthHandler 可以傳入不同的 bcrypeCost 參數，讓測試可以更快

一般使用時，bcryptCostDefault 是 10，但這樣測試要跑 30 秒。測試時改成 bcryptCostMin 只有 4，測試可以 2 秒跑完

由於 golang 沒有其他語言的 function(params=default) 這種預設參數值的語法糖，其中一種寫法就是這裡的 option pattern，用函數來當成 argument，我覺得是簡明又有擴展性的方式

觀察 sqlc 對於 db withTX 的處理，也是這種寫法的變形

richardlinio commented on internal/api/auth.go:
```diff
@@ -0,0 +1,185 @@
+	ctx, cancel := context.WithTimeout(c.Request.Context(), loginTimeout)
+	defer cancel()
+
+	user, err := h.authRepo.GetUserByUsername(ctx, req.Username)
+	if err != nil {
+		// Handle user not found (use dummy hash for timing attack prevention)
+		if errors.Is(err, repository.ErrUserNotFound) {
+			userExists = false
+			passwordHash = h.dummyPasswordHash
+		} else {
+			// All other errors are treated as internal server errors
+			// TODO: [HPB-211] Add proper logging here to capture the actual error for debugging
+			c.JSON(http.StatusInternalServerError, ErrorResponse{
+				Code:    http.StatusInternalServerError,
+				Message: "Internal server error",
+			})
+			return
+		}
+	} else {
+		userExists = true
+		passwordHash = user.Password
+	}
+
+	// Timing attack prevention: Always perform bcrypt comparison regardless of user existence.
+	// This ensures consistent response time whether user exists or not.
+	passwordErr := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(req.Password))
```
不管有沒有找到 username，都會等 hash 跑完後才回傳結果，避免 username not found 回傳時間比密碼錯誤更短所導致的時間差攻擊漏洞

richardlinio commented on internal/auth/jwt.go:
```diff
@@ -0,0 +1,117 @@
+
+	claims := &Claims{}
+	token, err := jwt.ParseWithClaims(tokenString, claims, func(token *jwt.Token) (any, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
```
檢驗演算法正確，避免演算法替換攻擊

richardlinio committed:
chore(middleware): add TODO comments for logging in auth error handling
- Add TODO comment to log errors when extracting Bearer token fails
- Add TODO comment to log errors when JWT token validation fails

richardlinio committed:
refactor(middleware): simplify HostHeaderValidation host check

- replace manual loop with slices.Contains for allowedHosts matching
- improve readability by using built-in slice utility function

richardlinio committed:
fix(middleware): improve HostHeaderValidation error handling and update tests
- Return 500 Internal Server Error if allowedHosts is empty instead of skipping validation
- Change error message from "Invalid host header" to "Invalid authorization request" on bad host header
- Add TODO comments for logging error details in HostHeaderValidation middleware
- Update tests to expect 500 status and error message when allowedHosts is empty
- Adjust test assertions to reflect new behavior of aborting request instead of calling next handler

JYu1999 commented on db/queries/auth.sql:
為什麼這個檔案是命名為 auth.sql 而不是 user.sql 呢？

是要用商業邏輯劃分 SQL 的意思嗎？

JYu1999 commented on db/queries/auth.sql:
```diff
@@ -0,0 +1,10 @@
+FROM
+    users
+WHERE
+    username = $1
+    AND deleted_at IS NULL
```
是否有考慮過 index 議題？

JYu1999 commented on scripts/seed.go:
1. Seeder 是否應該明確定義環境？有些 Seeder 是只能在開發環境執行的
2. 是因為什麼考量，最終決定將檔案放在 `/scripts` 底下呢？

JYu1999 commented on scripts/seed.go:
```diff
@@ -0,0 +1,85 @@
+		return fmt.Errorf("hash generation failed: %w", err)
+	}
+
+	query := `
+		INSERT INTO users (username, password, created_at, updated_at)
+		VALUES ($1, $2, NOW(), NOW())
+		ON CONFLICT (username) DO UPDATE
+			SET password = EXCLUDED.password,
+			    updated_at = NOW()
+	`
```
我們現在有使用 sqlc，seeder 為何不直接在 sqlc 建立一個「新增使用者」的 Query，而是要自己寫一個？

JYu1999 commented on internal/repository/auth.go:
這個跟 auth.sql 的問題有點像

repository 在設計上，通常會讓其和資料層是一致的。

例如 User 就會對應有一個 UserRepository

Repository 無論在命名上，或是裡面的商業邏輯，應該都只和取得資料相關，否則基本上就是在和 Service 「搶工作」

JYu1999 commented on internal/repository/auth.go:
```diff
@@ -0,0 +1,64 @@
+func (r *authRepository) GetUserByUsername(ctx context.Context, username string) (db.User, error) {
+	user, err := r.queries.GetUserByUsername(ctx, username)
+	if err != nil {
+		// Handle context errors
+		if errors.Is(err, context.Canceled) {
+			return db.User{}, ErrCanceled
+		}
+		if errors.Is(err, context.DeadlineExceeded) {
+			return db.User{}, ErrTimeout
+		}
+
+		// Handle database-specific errors
+		if errors.Is(err, pgx.ErrNoRows) {
+			return db.User{}, ErrUserNotFound
+		}
+
+		// Wrap other database errors
+		return db.User{}, fmt.Errorf("%w: %w", ErrDatabaseQuery, err)
```
我在想，如果之後新增了一個 `getUserById`，這邊錯誤處理是不是全部都要重寫？

或是在其他 repository 的查詢都需要重寫一次？

JYu1999 commented on internal/auth/jwt.go:
看起來這邊是屬於商業邏輯的部分，我覺得可以放在 `services/auth/`

JYu1999 commented on cmd/server/main.go:
```diff
@@ -38,13 +52,57 @@ func run() error {
 	// 初始化 Handler
 	handler := api.NewHandler(postStore)
 
+	// 從 .env 取得 JWT secret 並驗證
+	jwtSecret := os.Getenv("JWT_SECRET")
+	if err := auth.ValidateSecret(jwtSecret); err != nil {
+		return fmt.Errorf("invalid JWT secret: %w", err)
+	}
```
.env.example 和 .env.dev 應該需要修正？

JYu1999 replied:
漏看了拍謝

KR0924 commented on cmd/server/main.go:
```diff
@@ -7,15 +7,29 @@ import (
+	corsMaxAge = 24 * time.Hour
+)
+
+var (
```
DATABASE_URL, JWT_SECRET, ALLOWED_ORIGINS5 這幾個有沒有考慮放成 Config struct集中管理

KR0924 commented on cmd/server/main.go:
```diff
@@ -7,15 +7,29 @@ import (
+	errBackendPortNotSet         = errors.New("BACKEND_PORT environment variable must be set")
+)
 
 func run() error {
```
run function目前看起來好像做的事情有點多，後面如果隨著開發繼續可能會變得越來越大包，這邊可以先註記後續真的要改再回來重構

KR0924 commented on internal/api/auth.go:
```diff
@@ -0,0 +1,185 @@
+		return
+	}
+
+	c.JSON(http.StatusOK, LoginResponse{
+		Token: token,
+		User: LoginUser{
+			Username: user.Username,
+		},
+	})
```
雖然一開始有說只開這些資訊回傳，但我在想有沒有需要再多一個expires_time?

KR0924 commented on internal/auth/jwt.go:
```diff
@@ -0,0 +1,117 @@
+
+const (
+	// JWTExpiration is the JWT token expiration time
+	JWTExpiration = 7 * 24 * time.Hour // 7 days
+	// MinSecretLength is the minimum required length for JWT secret
+	MinSecretLength = 32
+	// BearerPrefix is the standard Bearer token prefix in Authorization header
+	BearerPrefix = "Bearer "
```
我如果要測試token過期後操作有沒有正常只能手動改這邊嗎?

KR0924 commented on internal/repository/auth.go:
```diff
@@ -0,0 +1,64 @@
+// NewAuthRepository creates a new AuthRepository instance
+func NewAuthRepository(queries AuthQuerier) AuthRepository {
+	return &authRepository{
+		queries: queries,
```
判斷query是否為nil(nil pointer dereference，雖然機率好像比較低但修改成本不高可以預防的話就先預防了)

richardlinio committed:
fix(auth): change status code for auth failures from 401 Unauthorized to 403 Forbidden

- update Login handler to return 403 Forbidden on invalid username or password
- update AuthRequired middleware to return 403 Forbidden for invalid authorization format and invalid token
- update tests in internal/api/auth_test.go and internal/middleware/auth_test.go to expect 403 Forbidden instead of 401 Unauthorized on authentication failures
